corehq/blobs/s3db.py

```diff
@@ -0,0 +1,162 @@
+        must be a seekable file-like object. NOTE: the value should not
+        be prefixed with `md5-` even though we store it that way.
+        :returns: A `BlobInfo` named tuple. The returned object has a
+        `name` member that must be used to get or delete the blob. It
+        should not be confused with the optional `basename` parameter.
+        """
+        name = self.get_unique_name(basename)
+        path = self.get_path(name, bucket)
+        if content_md5 is None:
+            params = {"body": content, "headers": {}}
+            calculate_md5(params)
+            content_md5 = params["headers"]["Content-MD5"]
+        with self.s3_bucket(True) as s3_bucket:
+            obj = s3_bucket.put_object(
+                Key=path,
+                Body=content,
+                ContentMD5=content_md5,
+            )
+        return BlobInfo(name, obj.content_length, "md5-" + content_md5)
+
+    def get(self, name, bucket=DEFAULT_BUCKET):
+        """Get a blob
+
+        :param name: The name of the object to get.
+        :param bucket: Optional bucket name. This must have the same
+        value that was passed to ``put``.
+        :raises: `NotFound` if the object does not exist.
+        :returns: A file-like object in binary read mode. The returned
+        object should be closed when finished reading.
+        """
+        path = self.get_path(name, bucket)
+        with self.s3_bucket() as s3_bucket:
+            try:
+                resp = s3_bucket.Object(path).get()
+            except ClientError as err:
+                if err.response["Error"]["Code"] == "NoSuchKey":
+                    raise NotFound(name, bucket)
+                raise
+            return ClosingContextProxy(resp["Body"])  # body stream
+
+    def delete(self, name=None, bucket=DEFAULT_BUCKET):
+        """Delete a blob
+
+        :param name: The name of the object to be deleted. The entire
+        bucket will be deleted if this is not specified.
+        :param bucket: Optional bucket name. This must have the same
+        value that was passed to ``put``.
+        :returns: True if the blob was deleted else false.
+        """
+        path = self.get_path(name, bucket)
+        success = True
+        with self.s3_bucket() as s3_bucket:
+            if name is None:
+                summaries = s3_bucket.objects.filter(Prefix=path + "/")
+                pages = ([{"Key": o.key} for o in page]
+                         for page in summaries.pages())
+            else:
+                pages = [[{"Key": path}]]
+            for objects in pages:
+                resp = s3_bucket.delete_objects(Delete={"Objects": objects})
+                if success:
+                    deleted = set(d["Key"] for d in resp.get("Deleted", []))
+                    success = all(o["Key"] in deleted for o in objects)
+        return success
+
+    @contextmanager
+    def s3_bucket(self, create=False):
```

Comment:
The `s3_bucket` members are not private because they are [accessed outside of the class ](https://github.com/dimagi/commcare-hq/blob/0c6d4ada2fe0bb16e6d660ffc32e7b9283d2e36c/corehq/blobs/tests/util.py#L62-L65)